submission/races.go

```diff
@@ -0,0 +1,227 @@
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		go func(i int, logURL string) {
+			time.Sleep(postInterval(i, parallelStart, PostBatchInterval))
+			if ctx.Err() != nil {
+				counter <- count{}
+				return
+			}
+			if firstRequested := state.request(logURL); !firstRequested {
+				counter <- count{}
+				return
+			}
+			if state.sumSCTsFor(group.LogURLs) >= group.MinInclusions {
+				completionMu.Lock()
+				if !earlyCompletion {
+					completion <- groupState{Name: group.Name, Success: true}
+					earlyCompletion = true
+				}
+				completionMu.Unlock()
+				counter <- count{}
+				return
+			}
+			sct, err := submitter.SubmitToLog(ctx, logURL, chain)
+			// TODO(Mercurrent): verify SCT
+			state.setResult(logURL, sct, err)
+			counter <- count{}
+		}(i, groupURLs[urlNum])
+	}
+	// Wait until either all logs within groups are processed or context is cancelled.
+	for i := 0; i < len(groupURLs); i++ {
+		select {
+		case <-ctx.Done():
+			break
+		case <-counter:
+		}
+	}
+	// Avoid signal duplicate.
+	if !earlyCompletion {
+		completion <- groupState{Name: group.Name, Success: state.sumSCTsFor(group.LogURLs) >= group.MinInclusions}
+	}
+}
+
+func parallelNums(groups ctpolicy.LogPolicyData) map[string]int {
+	nums := make(map[string]int)
+	var subsetSum int
+	for _, g := range groups {
+		nums[g.Name] = g.MinInclusions
+		if !g.IsBase {
+			subsetSum += g.MinInclusions
+		}
+	}
+	if _, hasBase := nums[ctpolicy.BaseName]; hasBase {
+		if nums[ctpolicy.BaseName] >= subsetSum {
+			nums[ctpolicy.BaseName] -= subsetSum
+		} else {
+			nums[ctpolicy.BaseName] = 0
+		}
+	}
+	return nums
+}
+
+// AssignedSCT represents SCT with logURL of log-producer.
+type AssignedSCT struct {
+	LogURL string
+	SCT    *ct.SignedCertificateTimestamp
+}
+
+func completenessError(groupComplete map[string]bool) error {
+	failedGroups := []string{}
+	for name, success := range groupComplete {
+		if !success {
+			failedGroups = append(failedGroups, name)
+		}
+	}
+	if len(failedGroups) > 0 {
+		return errors.New("Log-group(s) " + strings.Join(failedGroups, ", ") + "didn't receive enough SCTs.")
+	}
+	return nil
+}
+
+// GetSCTs picks required number of Logs according to policy-group logic and collects SCTs from them. Emits all collected SCTs even when any error produced.
+func GetSCTs(ctx context.Context, submitter Submitter, chain []ct.ASN1Cert, groups ctpolicy.LogPolicyData) ([]*AssignedSCT, error) {
+	parallelNums := parallelNums(groups)
+	submissions := safeSubmissionSet{results: make(map[string]*submissionResult)}
+	// channel listening to group-completion (failure) events
+	groupEvents := make(chan groupState, len(groups))
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, g := range groups {
+		go groupRace(subCtx, chain, g, parallelNums[g.Name], &submissions, groupEvents, submitter)
+	}
+
+	groupComplete := make(map[string]bool)
```

Comment:
Could do with a comment here, as this isn't just a set -- IIUC something like:
```golang
// A group name being present in the map indicates that group has
// completed processing, and the bool value indicates whether it
// completed successfully.
```

Reply:
Done.